File: apps/desktop/src/native/keyvault.ts
// 私鑰保管庫（B5，ADR-0053）：把 nsec 從「明文落在 localStorage」抽象成一層可換基質。
//
// - Tauri 執行期：走 OS 安全儲存（Windows Credential Manager / macOS Keychain /
//   Linux Secret Service），經 `key_set/key_get/key_delete` IPC（見 src-tauri keyvault.rs）。
// - 瀏覽器/開發：localStorage 後備，維持既有行為（非原生環境無 OS 金鑰庫）。
//
// 以 pubkey 為鍵，支援多身分（ADR-0045）。介面為 async——原生 IPC 天生非同步；
// 上層在開機時一次載入後交給同步引擎（見 App 開機流程）。

import { isWrapped } from "@cinder/core";
import { invoke, isTauri } from "@tauri-apps/api/core";

/** 私鑰保管庫：以 pubkey 存取該身分的 nsec。 */
export interface KeyVault {
  setKey(pubkey: string, nsec: string): Promise<void>;
  getKey(pubkey: string): Promise<string | null>;
  deleteKey(pubkey: string): Promise<void>;
}

const LS_PREFIX = "nb.key.";

/**
 * 瀏覽器後備（ADR-0112）：**nsec 絕不明文落盤**。
 *
 * 過去這裡直接 `localStorage.setItem(nsec)`——私鑰明文躺在磁碟上。那讓 ADR-0112 的儲存加密
 * 變成演戲（資料金鑰由 nsec 導出，而 nsec 就在旁邊）。
 *
 * 現在只接受 **Argon2id 密碼包裹的 blob**（`wrapSecret`）。收到明文 → **拒絕落盤**
 * ——這是強制點，不是建議：呼叫端若忘了包裹，資料不會被寫出去，而不是靜默地明文寫出去。
 * 不設密碼的使用者 → 不持久化，每次重新輸入 nsec（＝行動端現況，安全）。
 */
export const browserKeyVault: KeyVault = {
  async setKey(pubkey, nsec) {
    if (!isWrapped(nsec)) {
      // 明文私鑰不得落盤（ADR-0112 紅線）。順手清掉可能殘留的舊明文條目。
      try {
        localStorage.removeItem(LS_PREFIX + pubkey);
      } catch {
        /* 忽略 */
      }
      console.warn("[keyvault] 瀏覽器不接受明文 nsec 落盤（ADR-0112）；未設密碼＝不持久化");
      return;
    }
    try {
      localStorage.setItem(LS_PREFIX + pubkey, nsec);
    } catch {
      /* 配額/不可用時忽略 */
    }
  },
  async getKey(pubkey) {
    try {
      return localStorage.getItem(LS_PREFIX + pubkey);
    } catch {
      return null;
    }
  },
  async deleteKey(pubkey) {
    try {
      localStorage.removeItem(LS_PREFIX + pubkey);
    } catch {
      /* 忽略 */
    }
  },
};

/** Tauri：nsec 存 OS 安全儲存（keyring IPC）。 */
export const tauriKeyVault: KeyVault = {
  async setKey(pubkey, nsec) {
    await invoke("key_set", { pubkey, nsec });
  },
  async getKey(pubkey) {
    return (await invoke<string | null>("key_get", { pubkey })) ?? null;
  },
  async deleteKey(pubkey) {
    await invoke("key_delete", { pubkey });
  },
};

/** 依執行環境選用保管庫：Tauri → OS 金鑰庫；否則 localStorage 後備。 */
export function getKeyVault(): KeyVault {
  return isTauri() ? tauriKeyVault : browserKeyVault;
}

File: apps/desktop/src-tauri/src/keyvault.rs
//! OS 金鑰庫（B5，ADR-0053）：私鑰（nsec）託管於作業系統安全儲存
//! （Windows Credential Manager / macOS Keychain / Linux Secret Service），
//! **不再明文落地** localStorage/SQLite。需 `keyring` feature。
//!
//! 以固定 `SERVICE` + **pubkey 為帳號**存放，支援多身分（ADR-0045）——各身分的
//! nsec 互不覆蓋。純同步 API；上層（`#[tauri::command]`）再包成前端可 `invoke` 的橋。

use keyring::{Entry, Error as KeyringError};

/// 金鑰庫服務名稱（同一使用者下所有 Cinder 身分共用；以 pubkey 區分帳號）。
const SERVICE: &str = "app.cinder.desktop";

/// 金鑰庫存取失敗（包住後端錯誤字串，供 IPC 回傳前端）。
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct VaultError(pub String);

impl std::fmt::Display for VaultError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "金鑰庫錯誤：{}", self.0)
    }
}

impl std::error::Error for VaultError {}

fn entry(pubkey: &str) -> Result<Entry, VaultError> {
    Entry::new(SERVICE, pubkey).map_err(|e| VaultError(e.to_string()))
}

/// 存入（或覆寫）某身分（pubkey）的 nsec。
pub fn set_key(pubkey: &str, nsec: &str) -> Result<(), VaultError> {
    entry(pubkey)?.set_password(nsec).map_err(|e| VaultError(e.to_string()))
}

/// 取出某身分的 nsec；不存在回 `Ok(None)`。
pub fn get_key(pubkey: &str) -> Result<Option<String>, VaultError> {
    match entry(pubkey)?.get_password() {
        Ok(nsec) => Ok(Some(nsec)),
        Err(KeyringError::NoEntry) => Ok(None),
        Err(e) => Err(VaultError(e.to_string())),
    }
}

/// 刪除某身分的 nsec（登出/移除身分）；不存在視為成功（冪等）。
pub fn delete_key(pubkey: &str) -> Result<(), VaultError> {
    match entry(pubkey)?.delete_credential() {
        Ok(()) => Ok(()),
        Err(KeyringError::NoEntry) => Ok(()),
        Err(e) => Err(VaultError(e.to_string())),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    // 專用測試帳號，避免與真實身分/其他測試衝突；測前後皆清除。
    const TEST_PUBKEY: &str = "__cinder_keyvault_selftest__do_not_use__";

    #[test]
    fn set_get_delete_roundtrip_and_idempotent_delete() {
        let _ = delete_key(TEST_PUBKEY); // 清掉可能的殘留
        assert_eq!(get_key(TEST_PUBKEY).unwrap(), None, "初始應為空");

        set_key(TEST_PUBKEY, "nsec1_secret_original").unwrap();
        assert_eq!(get_key(TEST_PUBKEY).unwrap().as_deref(), Some("nsec1_secret_original"));

        // 覆寫（同一 pubkey 再存）
        set_key(TEST_PUBKEY, "nsec1_secret_rotated").unwrap();
        assert_eq!(get_key(TEST_PUBKEY).unwrap().as_deref(), Some("nsec1_secret_rotated"));

        delete_key(TEST_PUBKEY).unwrap();
        assert_eq!(get_key(TEST_PUBKEY).unwrap(), None, "刪除後應為空");
        delete_key(TEST_PUBKEY).unwrap(); // 重複刪除視為成功（冪等）
    }
}
